### ISSUE TITLE
AttributeName dynamic creation

### ISSUE THREAD

**Issue by keynmol** [scanamo/scanamo#222]
Hello. Is there a particular reason `AttributeName` is `private[scanamo]`? 

I'm looking for a nice way to support dynamic schemas - i.e. create update expressions from `List[Symbol] -> Value`, but access to the constructor is impossible without hacking with package names.

Is there a nice way to do it?

**Comment by regiskuckaertz** [comment]
Hello. Can't you just use `Symbol(yourDynamicallyGeneratedName) -> Value`? The conversion to `AttributeName` will then happen automatically.

**Comment by philwills** [comment]
I got a bit nervous after initially being a bit overly open with some scoping and this is possibly an over reaction the other way. I was initially a bit unsure of how it fit into the design, but am now pretty sure it's a good thing and should be used more widely, so I wouldn't object to exposing it now.

**Comment by keynmol** [comment]
@regiskuckaertz I might be missing something obvious, but without accessing `AttributeName` I can't trigger the implicit conversion in lists (because there's no derivation macro, probably):

```scala
@ set('attributes -> "hello")
res9: update.UpdateExpression = SimpleUpdateExpression(LeafSetExpression("updateSetattributes", Map("updateSetattributes" -> "attributes"), "update", {S: hello,}))

@ set(List('attributes, 'hello) -> "hello")
cmd10.sc:1: overloaded method value set with alternatives:
  [V](fieldValue: (com.gu.scanamo.query.AttributeName, V))(implicit evidence$16: com.gu.scanamo.DynamoFormat[V])com.gu.scanamo.update.UpdateExpression <and>
  (fields: (com.gu.scanamo.query.AttributeName, com.gu.scanamo.query.AttributeName))com.gu.scanamo.update.UpdateExpression
 cannot be applied to ((List[Symbol], String))
val res10 = set(List('attributes, 'hello) -> "hello")
            ^
Compilation Failed
```

And likewise I cannot do `... : List[AttributeName]` because it's inaccessible:

```scala
@ val l: List[com.gu.scanamo.query.AttributeName] = List('attributes, 'hello)
cmd6.sc:1: class AttributeName in package query cannot be accessed in package com.gu.scanamo.query
val l: List[com.gu.scanamo.query.AttributeName] = List('attributes, 'hello)
                                 ^
Compilation Failed
```

**Comment by regiskuckaertz** [comment]
Oh, I see what you mean. You're right, we need to do something about that 👍 

Not ideal, but in the meantime it might help to just do `set(Symbol(l.mkString(".")) -> "hello")`. `AttributeName` is just a convenience to make sure reserved DynamoDB keywords are properly escaped, so if you're careful it should be fine.

**Comment by regiskuckaertz** [comment]
Closed in favour of #389
